File: vlc-android/src/org/videolan/vlc/gui/audio/AudioBrowserFragment.java
/*****************************************************************************
 * AudioBrowserFragment.java
 *****************************************************************************
 * Copyright © 2011-2017 VLC authors and VideoLAN
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *****************************************************************************/

package org.videolan.vlc.gui.audio;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.tabs.TabLayout;

import org.videolan.medialibrary.media.MediaLibraryItem;
import org.videolan.medialibrary.media.MediaWrapper;
import org.videolan.vlc.MediaParsingServiceKt;
import org.videolan.vlc.R;
import org.videolan.vlc.gui.AudioPlayerContainerActivity;
import org.videolan.vlc.gui.PlaylistActivity;
import org.videolan.vlc.gui.SecondaryActivity;
import org.videolan.vlc.gui.preferences.PreferencesActivity;
import org.videolan.vlc.gui.view.FastScroller;
import org.videolan.vlc.gui.view.RecyclerSectionItemDecoration;
import org.videolan.vlc.gui.view.SwipeRefreshLayout;
import org.videolan.vlc.media.MediaUtils;
import org.videolan.vlc.util.Constants;
import org.videolan.vlc.util.Settings;
import org.videolan.vlc.util.WeakHandler;
import org.videolan.vlc.viewmodels.paged.MLPagedModel;
import org.videolan.vlc.viewmodels.paged.PagedAlbumsModel;
import org.videolan.vlc.viewmodels.paged.PagedArtistsModel;
import org.videolan.vlc.viewmodels.paged.PagedGenresModel;
import org.videolan.vlc.viewmodels.paged.PagedTracksModel;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class AudioBrowserFragment extends BaseAudioBrowser implements SwipeRefreshLayout.OnRefreshListener {
    public final static String TAG = "VLC/AudioBrowserFragment";

    private AudioBrowserAdapter mSongsAdapter;
    private AudioBrowserAdapter mArtistsAdapter;
    private AudioBrowserAdapter mAlbumsAdapter;
    private AudioBrowserAdapter mGenresAdapter;

    private PagedArtistsModel artistModel;
    private PagedAlbumsModel albumModel;
    private PagedTracksModel tracksModel;
    private PagedGenresModel genresModel;

    private TextView mEmptyView;
    private Button mMedialibrarySettingsBtn;
    private final RecyclerView[] mLists = new RecyclerView[MODE_TOTAL];
    private MLPagedModel<MediaLibraryItem>[] models;
    private SharedPreferences mSettings;
    private FastScroller mFastScroller;



    private static final String KEY_LISTS_POSITIONS = "key_lists_position";
    private static final int SET_REFRESHING = 103;
    private static final int UNSET_REFRESHING = 104;
    private static final int UPDATE_EMPTY_VIEW = 105;
    private final static int MODE_ARTIST = 0;
    private final static int MODE_ALBUM = 1;
    private final static int MODE_SONG = 2;
    private final static int MODE_GENRE = 3;
    private final static int MODE_TOTAL = 4; // Number of audio mProvider modes

    public final static String TAG_ITEM = "ML_ITEM";

    @Override
    protected boolean hasTabs() {
        return true;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (mSettings == null) mSettings = Settings.INSTANCE.getInstance(requireContext());
        if (models == null) setupModels();
        if (mSettings.getBoolean("audio_resume_card", true)) ((AudioPlayerContainerActivity)requireActivity()).proposeCard();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.audio_browser, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mEmptyView = view.findViewById(R.id.no_media);
        mMedialibrarySettingsBtn = view.findViewById(R.id.button_nomedia);
        mFastScroller = view.getRootView().findViewById(R.id.songs_fast_scroller);
        mFastScroller.attachToCoordinator((AppBarLayout) view.getRootView().findViewById(R.id.appbar), (CoordinatorLayout) view.getRootView().findViewById(R.id.coordinator), (FloatingActionButton) view.getRootView().findViewById(R.id.fab));
        mMedialibrarySettingsBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final FragmentActivity activity = requireActivity();
                final Intent intent = new Intent(activity.getApplicationContext(), SecondaryActivity.class);
                intent.putExtra("fragment", SecondaryActivity.STORAGE_BROWSER);
                startActivity(intent);
                activity.setResult(PreferencesActivity.RESULT_RESTART);
            }
        });
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        for (int i = 0; i < MODE_TOTAL; i++) mLists[i] = (RecyclerView) mViewPager.getChildAt(i);

        final String[] titles = new String[] {
                getString(R.string.artists),
                getString(R.string.albums),
                getString(R.string.tracks),
                getString(R.string.genres),
        };
        mViewPager.setOffscreenPageLimit(MODE_TOTAL - 1);
        mViewPager.setAdapter(new AudioPagerAdapter(mLists, titles));
        final int tabPosition = mSettings.getInt(Constants.KEY_AUDIO_CURRENT_TAB, 0);
        mViewPager.setCurrentItem(tabPosition);
        final ArrayList<Integer> positions = savedInstanceState != null ? savedInstanceState.getIntegerArrayList(KEY_LISTS_POSITIONS) : null;
        for (int i = 0; i< MODE_TOTAL; ++i) {
            final LinearLayoutManager llm = new LinearLayoutManager(getActivity());
            llm.setRecycleChildrenOnDetach(true);
            mLists[i].setLayoutManager(llm);
            mLists[i].setAdapter(mAdapters[i]);
            if (positions != null) mLists[i].scrollToPosition(positions.get(i));
            mLists[i].addOnScrollListener(mScrollListener);
            mLists[i].addItemDecoration(new RecyclerSectionItemDecoration(getResources().getDimensionPixelSize(R.dimen.recycler_section_header_height), true, models[i]));
        }
        mViewPager.setOnTouchListener(mSwipeFilter);
        mSwipeRefreshLayout.setOnRefreshListener(this);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        final ArrayList<Integer> positions = new ArrayList<>(MODE_TOTAL);
        for (int i = 0; i< MODE_TOTAL; ++i) {
            positions.add(((LinearLayoutManager)mLists[i].getLayoutManager()).findFirstCompletelyVisibleItemPosition());
        }
        outState.putIntegerArrayList(KEY_LISTS_POSITIONS, positions);
        super.onSaveInstanceState(outState);
    }

    private void setupModels() {
        final int current = mSettings.getInt(Constants.KEY_AUDIO_CURRENT_TAB, 0);
        for (int pass = 0 ; pass < 2; ++pass) {
            if ((pass != 0 ^ current == MODE_ARTIST) && mArtistsAdapter == null) {
                artistModel = ViewModelProviders.of(requireActivity(), new PagedArtistsModel.Factory(requireContext(), mSettings.getBoolean(Constants.KEY_ARTISTS_SHOW_ALL, false))).get(PagedArtistsModel.class);
                mArtistsAdapter = new AudioBrowserAdapter(MediaLibraryItem.TYPE_ARTIST, this);
            }
            if ((pass != 0 ^ current == MODE_ALBUM) && mAlbumsAdapter == null) {
                albumModel = ViewModelProviders.of(requireActivity(), new PagedAlbumsModel.Factory(requireContext(), null)).get(PagedAlbumsModel.class);
                mAlbumsAdapter = new AudioBrowserAdapter(MediaLibraryItem.TYPE_ALBUM, this);
            }
            if ((pass != 0 ^ current == MODE_SONG) && mSongsAdapter == null) {
                tracksModel = ViewModelProviders.of(requireActivity(), new PagedTracksModel.Factory(requireContext(), null)).get(PagedTracksModel.class);
                mSongsAdapter = new AudioBrowserAdapter(MediaLibraryItem.TYPE_MEDIA, this);
            }
            if ((pass != 0 ^ current == MODE_GENRE) && mGenresAdapter == null) {
                genresModel = ViewModelProviders.of(requireActivity(), new PagedGenresModel.Factory(requireContext())).get(PagedGenresModel.class);
                mGenresAdapter = new AudioBrowserAdapter(MediaLibraryItem.TYPE_GENRE, this);
            }
        }
        mAdapters = new AudioBrowserAdapter[] {mArtistsAdapter, mAlbumsAdapter, mSongsAdapter, mGenresAdapter};
        models = new MLPagedModel[] {artistModel, albumModel, tracksModel, genresModel};
        for (int pass = 0 ; pass < 2; ++pass) {
            for (int i = 0; i < models.length; ++i) {
                if (pass == 0 ^ current == i) continue;
                final int index = i;
                models[i].getPagedList().observe(this, new Observer<PagedList<MediaLibraryItem>>() {
                    @Override
                    public void onChanged(@Nullable PagedList<MediaLibraryItem> items) {
                        if (items != null) mAdapters[index].submitList(items);
                    }
                });
                models[i].getLoading().observe(this, new Observer<Boolean>() {
                    @Override
                    public void onChanged(@Nullable Boolean loading) {
                        if (loading == null || mViewPager.getCurrentItem() != index) return;
                        if (loading) mHandler.sendEmptyMessageDelayed(SET_REFRESHING, 300);
                        else mHandler.sendEmptyMessage(UNSET_REFRESHING);
                    }
                });
            }
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        setFabPlayShuffleAllVisibility();
        mFabPlay.setImageResource(R.drawable.ic_fab_shuffle);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.ml_menu_last_playlist).setVisible(true);
    }

    @Override
    public void onFabPlayClick(View view) {
        MediaUtils.INSTANCE.playAll(view.getContext(), tracksModel, 0, true);
    }

    private void setFabPlayShuffleAllVisibility() {
        setFabPlayVisibility(mSongsAdapter.getItemCount() > 2);
    }

    /**
     * Handle changes on the list
     */
    private final Handler mHandler = new AudioBrowserHandler(this);

    @Override
    public void onRefresh() {
        mActivity.closeSearchView();
        MediaParsingServiceKt.reloadLibrary(requireContext());
    }

    @Override
    public String getTitle() {
        return getString(R.string.audio);
    }

    @Override
    public boolean enableSearchOption() {
        return true;
    }

    private void updateEmptyView() {
        mEmptyView.setVisibility(getCurrentAdapter().isEmpty() ? View.VISIBLE : View.GONE);
        mMedialibrarySettingsBtn.setVisibility(getCurrentAdapter().isEmpty() ? View.VISIBLE : View.GONE);
        setFabPlayShuffleAllVisibility();
    }

    @Override
    public void onPageSelected(int position) {
        updateEmptyView();
        setFabPlayShuffleAllVisibility();
    }

    @Override
    public void onTabSelected(TabLayout.Tab tab) {
        super.onTabSelected(tab);
        mFastScroller.setRecyclerView(mLists[tab.getPosition()],models[tab.getPosition()]);
        mSettings.edit().putInt(Constants.KEY_AUDIO_CURRENT_TAB, tab.getPosition()).apply();
        final Boolean loading = getViewModel().getLoading().getValue();
        if (loading == null || !loading) mHandler.sendEmptyMessage(UNSET_REFRESHING);
        else mHandler.sendEmptyMessage(SET_REFRESHING);
    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {
        super.onTabUnselected(tab);
        onDestroyActionMode((AudioBrowserAdapter) mLists[tab.getPosition()].getAdapter());
        models[tab.getPosition()].restore();
    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {
        mLists[tab.getPosition()].smoothScrollToPosition(0);
    }

    @Override
    public void onClick(View v, int position, MediaLibraryItem item) {
        if (item == null) return;
        if (mActionMode != null) {
            super.onClick(v, position, item);
            return;
        }
        if (item.getItemType() == MediaLibraryItem.TYPE_MEDIA) {
            MediaUtils.INSTANCE.openMedia(getActivity(), (MediaWrapper) item);
            return;
        }
        final Intent i;
        switch (item.getItemType()) {
            case MediaLibraryItem.TYPE_ARTIST:
            case MediaLibraryItem.TYPE_GENRE:
                i = new Intent(getActivity(), SecondaryActivity.class);
                i.putExtra(SecondaryActivity.KEY_FRAGMENT, SecondaryActivity.ALBUMS_SONGS);
                i.putExtra(TAG_ITEM, item);
                break;
            case MediaLibraryItem.TYPE_ALBUM:
                i = new Intent(getActivity(), PlaylistActivity.class);
                i.putExtra(TAG_ITEM, item);
                break;
            default:
                return;
        }
        startActivity(i);
    }

    @Override
    public void onUpdateFinished(RecyclerView.Adapter adapter) {
        super.onUpdateFinished(adapter);
        if (adapter == getCurrentAdapter()) {
            mSwipeRefreshLayout.setEnabled(((LinearLayoutManager)getCurrentRV().getLayoutManager()).findFirstVisibleItemPosition() <= 0);
            updateEmptyView();
            mFastScroller.setRecyclerView(getCurrentRV(), getViewModel());
        } else setFabPlayShuffleAllVisibility();
    }

    @Override
    public MLPagedModel<MediaLibraryItem> getViewModel() {
        return models[mViewPager.getCurrentItem()];
    }

    @Override
    public AudioBrowserAdapter getCurrentAdapter() {
        return (AudioBrowserAdapter) (getCurrentRV()).getAdapter();
    }

    @Override
    protected RecyclerView getCurrentRV() {
        return mLists[mViewPager.getCurrentItem()];
    }

    private static class AudioBrowserHandler extends WeakHandler<AudioBrowserFragment> {
        AudioBrowserHandler(AudioBrowserFragment owner) {
            super(owner);
        }

        @Override
        public void handleMessage(Message msg) {
            final AudioBrowserFragment fragment = getOwner();
            if (fragment == null) return;
            switch (msg.what) {
                case SET_REFRESHING:
                    fragment.mSwipeRefreshLayout.setRefreshing(true);
                    break;
                case UNSET_REFRESHING:
                    removeMessages(SET_REFRESHING);
                    fragment.mSwipeRefreshLayout.setRefreshing(false);
                    break;
                case UPDATE_EMPTY_VIEW:
                    fragment.updateEmptyView();
            }
        }
    }

    public void updateArtists() {
        artistModel.showAll(mSettings.getBoolean(Constants.KEY_ARTISTS_SHOW_ALL, false));
        artistModel.refresh();
    }

    /*
     * Disable Swipe Refresh while scrolling horizontally
     */
    private View.OnTouchListener mSwipeFilter = new View.OnTouchListener() {
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            mSwipeRefreshLayout.setEnabled(event.getAction() == MotionEvent.ACTION_UP);
            return false;
        }
    };

    @Override
    public boolean allowedToExpand() {
        return getCurrentRV().getScrollState() == RecyclerView.SCROLL_STATE_IDLE;
    }
}

File: vlc-android/src/org/videolan/vlc/gui/audio/AudioBrowserFragment.kt
/*****************************************************************************
 * AudioBrowserFragment.java
 *
 * Copyright © 2011-2017 VLC authors and VideoLAN
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 */

package org.videolan.vlc.gui.audio

import android.content.Intent
import android.content.SharedPreferences
import android.content.res.Configuration
import android.os.Bundle
import android.os.Message
import android.util.SparseArray
import android.view.*
import android.widget.Button
import android.widget.TextView
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.lifecycle.Observer
import androidx.paging.PagedList
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.viewpager.widget.ViewPager
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.tabs.TabLayout
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.ObsoleteCoroutinesApi
import org.videolan.medialibrary.interfaces.media.AbstractMediaWrapper
import org.videolan.medialibrary.media.MediaLibraryItem
import org.videolan.vlc.R
import org.videolan.vlc.gui.*
import org.videolan.vlc.gui.view.FastScroller
import org.videolan.vlc.gui.view.RecyclerSectionItemDecoration
import org.videolan.vlc.gui.view.RecyclerSectionItemGridDecoration
import org.videolan.vlc.media.MediaUtils
import org.videolan.vlc.providers.medialibrary.MedialibraryProvider
import org.videolan.vlc.reloadLibrary
import org.videolan.vlc.util.*
import org.videolan.vlc.viewmodels.mobile.AudioBrowserViewModel
import org.videolan.vlc.viewmodels.mobile.getViewModel

@ObsoleteCoroutinesApi
@ExperimentalCoroutinesApi
class AudioBrowserFragment : BaseAudioBrowser<AudioBrowserViewModel>(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var songsAdapter: AudioBrowserAdapter
    private lateinit var artistsAdapter: AudioBrowserAdapter
    private lateinit var albumsAdapter: AudioBrowserAdapter
    private lateinit var genresAdapter: AudioBrowserAdapter

    private lateinit var emptyView: TextView
    private lateinit var medialibrarySettingsBtn: Button
    private val lists = mutableListOf<RecyclerView>()
    private lateinit var settings: SharedPreferences
    private lateinit var fastScroller: FastScroller
    override val hasTabs = true
    private var spacing = 0
    private var restorePositions: SparseArray<Int> = SparseArray()

    /**
     * Handle changes on the list
     */
    private val handler = AudioBrowserHandler(this)

    /*
     * Disable Swipe Refresh while scrolling horizontally
     */
    private val swipeFilter = View.OnTouchListener { _, event ->
        swipeRefreshLayout.isEnabled = event.action == MotionEvent.ACTION_UP
        false
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        spacing = requireActivity().resources.getDimension(R.dimen.kl_small).toInt()


        if (!::settings.isInitialized) settings = Settings.getInstance(requireContext())
        if (!::songsAdapter.isInitialized) setupModels()
        if (viewModel.showResumeCard) {
            (requireActivity() as AudioPlayerContainerActivity).proposeCard()
            viewModel.showResumeCard = false
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.audio_browser, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        emptyView = view.findViewById(R.id.no_media)
        medialibrarySettingsBtn = view.findViewById(R.id.button_nomedia)
        fastScroller = view.rootView.findViewById(R.id.songs_fast_scroller)
        fastScroller.attachToCoordinator(view.rootView.findViewById<View>(R.id.appbar) as AppBarLayout, view.rootView.findViewById<View>(R.id.coordinator) as CoordinatorLayout, view.rootView.findViewById<View>(R.id.fab) as FloatingActionButton)
        medialibrarySettingsBtn.setOnClickListener {
            activity?.run {
                val intent = Intent(applicationContext, SecondaryActivity::class.java)
                intent.putExtra("fragment", SecondaryActivity.STORAGE_BROWSER)
                startActivity(intent)
                setResult(RESULT_RESTART)
            }
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewPager.let { viewPager ->
            val views = ArrayList<View>(MODE_TOTAL)
            for (i in 0 until MODE_TOTAL) {
                views.add(viewPager.getChildAt(i))
                lists.add(viewPager.getChildAt(i).findViewById(R.id.audio_list))
            }
            val titles = arrayOf(getString(R.string.artists), getString(R.string.albums), getString(R.string.tracks), getString(R.string.genres))
            viewPager.offscreenPageLimit = MODE_TOTAL - 1
            val audioPagerAdapter = AudioPagerAdapter(views.toTypedArray(), titles)
            @Suppress("UNCHECKED_CAST")
            viewPager.adapter = audioPagerAdapter
            val tabPosition = settings.getInt(KEY_AUDIO_CURRENT_TAB, 0)
            currentTab = tabPosition
            val position = savedInstanceState?.getIntegerArrayList(KEY_LISTS_POSITIONS)
            position?.withIndex()?.forEach {
                restorePositions.put(it.index, it.value)
            }
            for (i in 0 until MODE_TOTAL) {
                setupLayoutManager(i)
                (lists[i].layoutManager as LinearLayoutManager).recycleChildrenOnDetach = true
                val list = lists[i]
                list.adapter = adapters[i]
                list.addOnScrollListener(scrollListener)
                if (viewModel.providersInCard[i]) {
                    val lp = list.layoutParams
                    lp.width = ViewGroup.LayoutParams.MATCH_PARENT
                    list.layoutParams = lp
                }
            }
            viewPager.setOnTouchListener(swipeFilter)
            swipeRefreshLayout.setOnRefreshListener(this)
            viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
                override fun onPageScrollStateChanged(state: Int) {}

                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

                override fun onPageSelected(position: Int) {
                    activity?.invalidateOptionsMenu()
                }
            })
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)

        val itemSize = RecyclerSectionItemGridDecoration.getItemSize(requireActivity().getScreenWidth(), nbColumns, spacing)
        for (i in 0 until MODE_TOTAL) {
            if (lists[i].layoutManager is GridLayoutManager) {
                val gridLayoutManager = lists[i].layoutManager as GridLayoutManager
                gridLayoutManager.spanCount = nbColumns
                lists[i].layoutManager = gridLayoutManager
                adapters[i].cardSize = itemSize
                adapters[i].notifyItemRangeChanged(gridLayoutManager.findFirstVisibleItemPosition(), gridLayoutManager.findLastVisibleItemPosition() - gridLayoutManager.findFirstVisibleItemPosition())
            }
        }
    }

    private fun setupLayoutManager(index: Int) {
        if (lists[index].itemDecorationCount > 0) {
            lists[index].removeItemDecorationAt(0)
        }
        when (viewModel.providersInCard[index]) {
            true -> {
                adapters[index].cardSize = RecyclerSectionItemGridDecoration.getItemSize(requireActivity().getScreenWidth(), nbColumns, spacing)
                displayListInGrid(lists[index], adapters[index], viewModel.providers[index] as MedialibraryProvider<MediaLibraryItem>, spacing)
            }
            else -> {
                adapters[index].cardSize = -1
                lists[index].addItemDecoration(RecyclerSectionItemDecoration(resources.getDimensionPixelSize(R.dimen.recycler_section_header_height), true, viewModel.providers[index]))
                lists[index].layoutManager = LinearLayoutManager(activity)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        val positions = ArrayList<Int>(MODE_TOTAL)
        for (i in 0 until MODE_TOTAL) {
            positions.add((lists[i].layoutManager as LinearLayoutManager).findFirstCompletelyVisibleItemPosition())
        }
        outState.putIntegerArrayList(KEY_LISTS_POSITIONS, positions)
        super.onSaveInstanceState(outState)
    }

    private fun setupModels() {
        viewModel = getViewModel()

        val itemSize = RecyclerSectionItemGridDecoration.getItemSize(requireActivity().getScreenWidth(), nbColumns, spacing)

        artistsAdapter = AudioBrowserAdapter(MediaLibraryItem.TYPE_ARTIST, this, cardSize = if (viewModel.providersInCard[0]) itemSize else -1)
        albumsAdapter = AudioBrowserAdapter(MediaLibraryItem.TYPE_ALBUM, this, cardSize = if (viewModel.providersInCard[1]) itemSize else -1)
        songsAdapter = AudioBrowserAdapter(MediaLibraryItem.TYPE_MEDIA, this, cardSize = if (viewModel.providersInCard[2]) itemSize else -1)
        genresAdapter = AudioBrowserAdapter(MediaLibraryItem.TYPE_GENRE, this)
        adapters = arrayOf(artistsAdapter, albumsAdapter, songsAdapter, genresAdapter)
        for ((index, provider) in viewModel.providers.withIndex()) {
            provider.pagedList.observe(this, Observer { items ->
                @Suppress("UNCHECKED_CAST")
                if (items != null) adapters[index].submitList(items as PagedList<MediaLibraryItem>?)
                restorePositions.get(index)?.let {
                    lists[index].scrollToPosition(it)
                    restorePositions.delete(index)
                }
            })
            provider.loading.observe(this, Observer { loading ->
                if (loading == null || currentTab != index) return@Observer
                if (loading)
                    handler.sendEmptyMessageDelayed(SET_REFRESHING, 300)
                else
                    handler.sendEmptyMessage(UNSET_REFRESHING)

                (activity as? MainActivity)?.refreshing = loading
                updateEmptyView()
            })
        }
    }

    override fun onStart() {
        super.onStart()
        setFabPlayShuffleAllVisibility()
        fabPlay?.setImageResource(R.drawable.ic_fab_shuffle)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        menu.findItem(R.id.ml_menu_last_playlist)?.isVisible = true
        (viewModel.providers[currentTab]).run {
            menu.findItem(R.id.ml_menu_sortby).isVisible = canSortByName()
            menu.findItem(R.id.ml_menu_sortby_filename).isVisible = canSortByFileNameName()
            menu.findItem(R.id.ml_menu_sortby_artist_name).isVisible = canSortByArtist()
            menu.findItem(R.id.ml_menu_sortby_album_name).isVisible = canSortByAlbum()
            menu.findItem(R.id.ml_menu_sortby_length).isVisible = canSortByDuration()
            menu.findItem(R.id.ml_menu_sortby_date).isVisible = canSortByReleaseDate()
            menu.findItem(R.id.ml_menu_sortby_last_modified).isVisible = canSortByLastModified()
            menu.findItem(R.id.ml_menu_sortby_number).isVisible = false
            menu.findItem(R.id.ml_menu_display_grid).isVisible = currentTab in 0..2 && !viewModel.providersInCard[currentTab]
            menu.findItem(R.id.ml_menu_display_list).isVisible = currentTab in 0..2 && viewModel.providersInCard[currentTab]
        }
        sortMenuTitles()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.ml_menu_display_list, R.id.ml_menu_display_grid -> {
                viewModel.providersInCard[currentTab] = item.itemId == R.id.ml_menu_display_grid
                setupLayoutManager(currentTab)
                lists[currentTab].adapter = adapters[currentTab]
                activity?.invalidateOptionsMenu()
                Settings.getInstance(requireActivity()).edit().putBoolean(viewModel.displayModeKeys[currentTab], item.itemId == R.id.ml_menu_display_grid).apply()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun sortBy(sort: Int) {
        viewModel.providers[currentTab].sort(sort)
    }

    override fun onFabPlayClick(view: View) {
        MediaUtils.playAll(view.context, viewModel.tracksProvider, 0, true)
    }

    private fun setFabPlayShuffleAllVisibility() {
        setFabPlayVisibility(songsAdapter.itemCount > 2)
    }

    override fun onRefresh() {
        (requireActivity() as ContentActivity).closeSearchView()
        requireContext().reloadLibrary()
        viewModel.setLoading()
    }

    override fun getTitle(): String = getString(R.string.audio)

    override fun enableSearchOption(): Boolean {
        return true
    }

    private fun updateEmptyView() {
        val emptyVisibility = getViewModel().providers[currentTab].isEmpty() && getViewModel().providers[currentTab].loading.value == false
        emptyView.visibility = if (emptyVisibility) View.VISIBLE else View.GONE
        medialibrarySettingsBtn.visibility = if (emptyVisibility) View.VISIBLE else View.GONE
        setFabPlayShuffleAllVisibility()
    }

    override fun onPageSelected(position: Int) {
        updateEmptyView()
        setFabPlayShuffleAllVisibility()
    }

    override fun onTabSelected(tab: TabLayout.Tab) {
        super.onTabSelected(tab)
        fastScroller.setRecyclerView(lists[tab.position], viewModel.providers[tab.position])
        settings.edit().putInt(KEY_AUDIO_CURRENT_TAB, tab.position).apply()
        if (viewModel.providers[currentTab].isRefreshing)
            handler.sendEmptyMessage(SET_REFRESHING)
        else
            handler.sendEmptyMessage(UNSET_REFRESHING)
        activity?.invalidateOptionsMenu()
    }

    override fun onTabUnselected(tab: TabLayout.Tab) {
        super.onTabUnselected(tab)
        onDestroyActionMode(lists[tab.position].adapter as AudioBrowserAdapter?)
        viewModel.restore()
    }

    override fun onTabReselected(tab: TabLayout.Tab) {
        lists[tab.position].smoothScrollToPosition(0)
    }

    override fun onCtxAction(position: Int, option: Int) {
        @Suppress("UNCHECKED_CAST")
        if (option == CTX_PLAY_ALL) MediaUtils.playAll(requireContext(), viewModel.providers[currentTab] as MedialibraryProvider<AbstractMediaWrapper>, position, false)
        else super.onCtxAction(position, option)
    }

    override fun onClick(v: View, position: Int, item: MediaLibraryItem) {
        if (actionMode != null) {
            super.onClick(v, position, item)
            return
        }
        if (item.itemType == MediaLibraryItem.TYPE_MEDIA) {
            MediaUtils.openMedia(activity, item as AbstractMediaWrapper)
            return
        }
        val i: Intent
        when (item.itemType) {
            MediaLibraryItem.TYPE_ARTIST, MediaLibraryItem.TYPE_GENRE -> {
                i = Intent(activity, SecondaryActivity::class.java)
                i.putExtra(SecondaryActivity.KEY_FRAGMENT, SecondaryActivity.ALBUMS_SONGS)
                i.putExtra(TAG_ITEM, item)
            }
            MediaLibraryItem.TYPE_ALBUM -> {
                i = Intent(activity, PlaylistActivity::class.java)
                i.putExtra(TAG_ITEM, item)
            }
            else -> return
        }
        startActivity(i)
    }

    override fun onUpdateFinished(adapter: RecyclerView.Adapter<*>) {
        super.onUpdateFinished(adapter)
        if (adapter === getCurrentAdapter()) {
            swipeRefreshLayout.isEnabled = (getCurrentRV().layoutManager as LinearLayoutManager).findFirstVisibleItemPosition() <= 0
            updateEmptyView()
            fastScroller.setRecyclerView(getCurrentRV(), viewModel.providers[currentTab])
        } else
            setFabPlayShuffleAllVisibility()
    }

    override fun getCurrentRV() = lists[currentTab]

    override fun getCurrentAdapter() = adapters[currentTab]

    private class AudioBrowserHandler internal constructor(owner: AudioBrowserFragment) : WeakHandler<AudioBrowserFragment>(owner) {
        override fun handleMessage(msg: Message) {
            val fragment = owner ?: return
            when (msg.what) {
                SET_REFRESHING -> fragment.swipeRefreshLayout.isRefreshing = true
                UNSET_REFRESHING -> {
                    removeMessages(SET_REFRESHING)
                    fragment.swipeRefreshLayout.isRefreshing = false
                }
                UPDATE_EMPTY_VIEW -> fragment.updateEmptyView()
            }
        }
    }

    override fun allowedToExpand() = getCurrentRV().scrollState == RecyclerView.SCROLL_STATE_IDLE

    companion object {
        val TAG = "VLC/AudioBrowserFragment"

        private const val KEY_LISTS_POSITIONS = "key_lists_position"
        private const val SET_REFRESHING = 103
        private const val UNSET_REFRESHING = 104
        private const val UPDATE_EMPTY_VIEW = 105
        private const val MODE_TOTAL = 4 // Number of audio lists

        const val TAG_ITEM = "ML_ITEM"
    }
}
